### ISSUE TITLE
WebSocket Cloned Object not being freed because of exception.

### ISSUE THREAD

**Issue by Coldzer0** [synopse/mORMot2#95]
While testing some cases, I got this memory leak using Eurekalog.

Here in this code
https://github.com/synopse/mORMot2/blob/18c35b7c4094f79ef3e4f249ba63718bfd609dd1/src/net/mormot.net.ws.async.pas#L256-L273

As you can see, there's a rise of an exception after cloning the protocol object and before freeing it.

![image](https://user-images.githubusercontent.com/6647006/172987366-064ea1c5-d61e-448f-a4c6-6bc80d660bfd.png)

**Comment by synopse** [comment]
Nice catch!

Should be fixed now.

**Comment by synopse** [comment]
In fact, the exception was overkill here.
It will now return 404 with https://github.com/synopse/mORMot2/commit/e3223fd6
as the blocking server does.

**Comment by Coldzer0** [comment]
> In fact, the exception was overkill here.

That's true Because after testing the free fix here https://github.com/synopse/mORMot2/commit/bfc4422bf9d86452ce0a4e10dd6aa3441d39e2c9 
Another Access Violation started, And I was about to report it.

I'll check the https://github.com/synopse/mORMot2/commit/e3223fd6 and report back after testing.

<br><br>

btw I was testing the download of a file using the `STATICFILE_CONTENT_TYPE` flag like this

```pascal
if Ctxt.Url = '/installer.zip' then
begin
  Ctxt.OutContentType := STATICFILE_CONTENT_TYPE;
  Ctxt.OutContent := 'G:\installers\latest_stable.zip';
end;
```

But the server is sending data at maximum speed like 700 MB/S (in my case) 
How can I control the speed of the download? The `MaximumAllowedContentLength` didn't do anything in this case.

**Comment by synopse** [comment]
Why do you need to cut down the download speed?
On production, the available bandwidth will be shared among all connections: the async server is sending all output at the pace of each buffer.

**Comment by Coldzer0** [comment]
For the download speed, I have a business logic in the server related to the users to limit the download speed or make it faster.

Plus the upload part, I didn't find any way in mormot2 to handle file uploads from an HTTP client.

The thing Is, I have a Client that needs to upload multiple files simultaneously, 
but he wants to limit the Speed because his internet is not that fast.

And I found out that the MaximumAllowedContentLength is only for the HTTP Server.

Maybe I might use Nginx as a proxy to handle the speed part! I'll check it out later.

By the way, the Speed is super impressive, especially since it didn't use any memory from the App part.
![image](https://user-images.githubusercontent.com/6647006/173019420-34e92477-0b71-40ba-b1ae-ca0a2e98144e.png)

**Comment by synopse** [comment]
So you expect the download speed to be limited depending on each connection/user?
It won't be feasible with nginx.

Note that for nginx, you can set NginxSendFileFrom() to let the nginx proxy do the work.
On Linux, it would use sendfile() or tlssendfile() to make it even more efficient.

About speed and memory, it just serves a TFileStream content directly into the socket connection buffers, with async sending.
So it actually minimize resources usage.

About the upload, direct file support is in my to-do list.

Note: all this discussion would benefit from being in our forum, because it would stay available, whereas here it is "lost" once the issue is closed.

**Comment by Coldzer0** [comment]
Thanks for the reply.
I'll open a new thread in the forum about the current discussion.

And the last fix works very well, And the ProcessClass is very helpful.

Thanks for your fantastic work, as usual.
